service/history/callback_queue_active_task_executor.go

```diff
@@ -0,0 +1,293 @@
+		task.GetNamespaceID(),
+	)
+	metricsTags := []metrics.Tag{
+		namespaceTag,
+		metrics.TaskTypeTag(taskType),
+	}
+
+	// Check namespace handover state once when processing is started. Callback tasks may take up to 10 seconds
+	// by default but we avoid checking again before committing the result to complete inflight tasks even if
+	// the namespace transitions to handover state during processing.
+	if replicationState == enumspb.REPLICATION_STATE_HANDOVER {
+		// TODO: Move this logic to queues.Executable when metrics tags don't need to be returned from task executor.
+		return queues.ExecuteResponse{
+			ExecutionMetricTags: metricsTags,
+			ExecutedAsActive:    true,
+			ExecutionErr:        consts.ErrNamespaceHandover,
+		}
+	}
+
+	var err error
+	switch task := task.(type) {
+	case *tasks.CallbackTask:
+		err = t.processCallbackTask(ctx, task)
+	default:
+		err = errUnknownTransferTask
+	}
+
+	return queues.ExecuteResponse{
+		ExecutionMetricTags: metricsTags,
+		ExecutedAsActive:    true,
+		ExecutionErr:        err,
+	}
+}
+
+func (t *callbackQueueActiveTaskExecutor) processCallbackTask(ctx context.Context, task *tasks.CallbackTask) error {
+	ctx, cancel := context.WithTimeout(ctx, t.config.CallbackTaskTimeout())
+	defer cancel()
+
+	var completion nexus.OperationCompletion
+	var url string
+	err := t.accessCallback(ctx, task, func(wfCtx workflow.Context, ms workflow.MutableState, callback *persistencespb.CallbackInfo) error {
+		switch variant := callback.PublicInfo.GetCallback().GetVariant().(type) {
+		case *commonpb.Callback_Nexus_:
+			url = variant.Nexus.GetUrl()
+			var err error
+			completion, err = t.getNexusCompletion(ctx, ms)
+			return err
+		default:
+			return serviceerror.NewUnimplemented(fmt.Sprintf("unprocessable callback variant: %v", variant))
+		}
+	})
+	if err != nil {
+		return err
+	}
+	return t.processNexusCallbackTask(ctx, task, url, completion)
+}
+
+// getNexusCompletion converts a workflow completion event into a [nexus.OperationCompletion].
+// Completions may be sent to arbitrary third parties and we intentionally do not include any termination reasons and
+// expose only failure messages.
+func (t *callbackQueueActiveTaskExecutor) getNexusCompletion(ctx context.Context, ms workflow.MutableState) (nexus.OperationCompletion, error) {
+	ce, err := ms.GetCompletionEvent(ctx)
+	if err != nil {
+		return nil, err
+	}
+	switch ce.GetEventType() {
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED:
+		payloads := ce.GetWorkflowExecutionCompletedEventAttributes().GetResult().GetPayloads()
+		var payload *commonpb.Payload
+		if len(payloads) > 0 {
+			payload = payloads[0]
+		} else {
+			payload = &commonpb.Payload{}
+		}
+		completion, err := nexus.NewOperationCompletionSuccessful(payload, nexus.OperationCompletionSuccesfulOptions{
+			Serializer: t.payloadSerializer,
+		})
+		if err != nil {
+			return nil, serviceerror.NewInternal(fmt.Sprintf("failed to construct Nexus completion: %v", err))
```

Comment:
should not you use %w here?

Reply:
Hmm.. yeah perhaps I'll wrap the internal error along with `err`.
We tend to construct serviceerrors with these constructors, so I went with that approach.
`fmt.Sprintf` doesn't support `%w` wrapping.